### PR TITLE
Initial filemod support monitoring mkdir calls

### PR DESCRIPTION
**Pull request by pinealservo**
This contains a lot of relatively minor refactoring to make some code written for process events also work for file events. Some more re-organization to move the shared code to a better location is coming.

The file event code is just a starting point, but an event type is in place and it successfully fills out create events when mkdir is used.

### CODE REVIEW

File: src/file-events.c [hunk 1]
```diff
@@ -0,0 +1,224 @@
...
+
+static __always_inline void store_dentry(struct pt_regs *ctx, void *dentry)
+{
+    u64 pid_tgid = bpf_get_current_pid_tgid();
+    error_info_t info = {0};
+
+    load_event(incomplete_mkdirs, pid_tgid, incomplete_mkdir_t);
+    event.target_dentry = dentry;
+    bpf_map_update_elem(&incomplete_mkdirs, &pid_tgid, &event, BPF_ANY);
+
```
Comment (rc-vbundage) [inline src/file-events.c:139]: Add a return!

File: src/file-events.c [hunk 2]
```diff
@@ -0,0 +1,224 @@
...
+    EventMismatch:
+    info.stored_pid_tgid = event.pid_tgid;
+    set_local_warning(W_PID_TGID_MISMATCH, info);
```
Comment (rc-vbundage) [inline src/file-events.c:142]: We'll want to a push a warning here.

File: src/file-events.c [hunk 3]
```diff
@@ -0,0 +1,224 @@
...
+    return;
+}
+
+static __always_inline void store_dentry(struct pt_regs *ctx, void *dentry)
+{
+    u64 pid_tgid = bpf_get_current_pid_tgid();
+    error_info_t info = {0};
+
+    load_event(incomplete_mkdirs, pid_tgid, incomplete_mkdir_t);
+    event.target_dentry = dentry;
```
Comment (rc-vbundage) [inline src/file-events.c:137]: we should check that this is NULL to prevent overwriting when using overlayfs

File: src/file-events.c [hunk 4]
```diff
@@ -0,0 +1,232 @@
...
+#include "bpf_helpers.h"
+#include "types.h"
+#include "offsets.h"
+#include "common.h"
+#include "process/helpers.h"
+#include "process/buffer.h"
+#include "process/path.h"
+#include "process/warning.h"
+
+// The map where process event messages get emitted to
```
Comment (rc-vbundage) [inline src/file-events.c:15]:
```suggestion
// The map where file event messages get emitted to
```

File: src/file-events.c [hunk 5]
```diff
@@ -0,0 +1,232 @@
...
+    ResolveName:
+    ret = write_path(ctx, cached_path, &cursor, tail_call);
+    cached_path->next_dentry = NULL;
+    if (ret < 0) goto EmitWarning;
+
+    write_null_char(cursor.buffer, cursor.offset);
+    push_flexible_file_message(ctx, fm, *cursor.offset);
+    return;
+
+    EventMismatch:
```
Comment (rc-vbundage) [inline src/file-events.c:195]: question: Does this need to be `EventMismatch:;` (end with semicolon)?

Comment (pinealservo) [reply src/file-events.c:212]: It's not normal to put a semicolon after a label; I think it was done in some places to allow a variable declaration immediately after the label, which isn't normally allowed.

**Review by rc-andres — APPROVED** [review]

**Pull request merged**